File: conformance/code/employee/src/main/java/com/digit/employee/pubsub/MigrationConsumer.java
package com.digit.employee.pubsub;

import com.digit.employee.config.EmployeeProperties;
import com.digit.tenant.migration.MigrationService;
import org.digit.tracer.pubsub.PubSubClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

import java.nio.charset.StandardCharsets;

/**
 * Consumes tenant-migration events and triggers schema creation/migration.
 * Mirrors Go internal/pubsub/migration_consumer.go (works with both Kafka and Redis via tracer/pubsub).
 *
 * <p>Subscribes via the official tracer {@link PubSubClient}. When no client is configured
 * (digit.tracer.pubsub.type unset) consumption is a graceful no-op and manual migration via the
 * /internal/migrate API is still available.
 */
@Component
public class MigrationConsumer {

    private static final Logger log = LoggerFactory.getLogger(MigrationConsumer.class);

    private final PubSubClient pubSubClient; // may be null
    private final EmployeeProperties props;
    private final MigrationService migrationService;

    @Autowired
    public MigrationConsumer(@Autowired(required = false) PubSubClient pubSubClient,
                             EmployeeProperties props,
                             MigrationService migrationService) {
        this.pubSubClient = pubSubClient;
        this.props = props;
        this.migrationService = migrationService;
    }

    /** Starts the consumer in the background once the app is ready (Go: go consumer.StartWithRetry). */
    @EventListener(ApplicationReadyEvent.class)
    public void onApplicationReady() {
        Thread t = new Thread(this::startWithRetry, "migration-consumer");
        t.setDaemon(true);
        t.start();
    }

    /** Begins consuming migration events. Mirrors Go Start gating. */
    public void start() throws Exception {
        if (!props.getTenantMigration().isEnabled()) {
            log.info("Migration consumer disabled: schema separation not enabled");
            return;
        }
        if (pubSubClient == null) {
            log.info("Migration consumer disabled: PubSub not available "
                    + "(use /internal/migrate API for manual migrations)");
            return;
        }
        if (migrationService == null) {
            log.info("Migration consumer disabled: migration service not initialized");
            return;
        }
        String topic = props.getTenantMigration().getTopic();
        String consumerGroup = "kafka".equals(props.getPubsub().getType())
                ? props.getPubsub().getKafka().getConsumerGroup()
                : props.getPubsub().getRedis().getConsumerGroup();
        log.info("Starting migration consumer on topic: {} (type: {})", topic, props.getPubsub().getType());
        pubSubClient.subscribe(topic, consumerGroup, this::handleMigrationEvent);
    }

    /** Retries Start on failure, mirroring Go StartWithRetry. */
    public void startWithRetry() {
        while (true) {
            try {
                start();
                return;
            } catch (Exception e) {
                log.error("Migration consumer error, retrying in 2 seconds", e);
                try {
                    Thread.sleep(2000);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    return;
                }
            }
        }
    }

    private void handleMigrationEvent(byte[] payload) {
        if (payload == null) {
            return;
        }
        log.info("[MIGRATION_EVENT] Received event: {}", new String(payload, StandardCharsets.UTF_8));
        try {
            migrationService.handleMessage(payload);
        } catch (Exception e) {
            log.error("Failed to handle migration event, payload={}",
                    new String(payload, StandardCharsets.UTF_8), e);
        }
    }
}

File: conformance/code/employee-go/internal/pubsub/migration_consumer.go
package pubsub

import (
	"context"
	"employee/internal/config"
	"encoding/json"
	"log"
	"time"

	tenantmigration "github.com/digitnxt/digit3/src/libraries/tenant-migration"

	tracerobs "github.com/digitnxt/digit3/src/libraries/tracer/observability"
	tracerpubsub "github.com/digitnxt/digit3/src/libraries/tracer/pubsub"
)

// MigrationConsumer consumes migration events and triggers schema creation
// Supports both Kafka and Redis Streams via tracer/pubsub
type MigrationConsumer struct {
	pubsubClient tracerpubsub.PubSubClient
	config       *config.Config
	logger       *tracerobs.OTelLogger
	migrationSvc *tenantmigration.Service
}

// NewMigrationConsumer creates a new migration consumer
func NewMigrationConsumer(
	pubsubClient tracerpubsub.PubSubClient,
	cfg *config.Config,
	migrationSvc *tenantmigration.Service,
) *MigrationConsumer {
	return &MigrationConsumer{
		pubsubClient: pubsubClient,
		config:       cfg,
		logger:       tracerobs.GetOTelLogger(),
		migrationSvc: migrationSvc,
	}
}

// Start begins consuming migration events
// Works with both Kafka and Redis Streams (configured via tracer/pubsub)
func (mc *MigrationConsumer) Start(ctx context.Context) error {
	// Migration consumer requires:
	// 1. Schema separation enabled
	// 2. PubSub client available (Kafka or Redis)
	// 3. Migration service initialized
	// Note: Migration service can work without PubSub via /internal/migrate API
	if !mc.config.TenantMigration.Enabled {
		log.Printf("Migration consumer disabled: schema separation not enabled")
		return nil
	}

	if mc.pubsubClient == nil {
		log.Printf("Migration consumer disabled: PubSub not available (use /internal/migrate API for manual migrations)")
		return nil
	}

	if mc.migrationSvc == nil {
		log.Printf("Migration consumer disabled: migration service not initialized")
		return nil
	}

	pubsubType := mc.config.PubSub.Type
	if pubsubType == "" {
		pubsubType = "kafka"
	}

	log.Printf("Starting migration consumer on topic: %s (type: %s)", mc.config.TenantMigration.Topic, pubsubType)

	// Subscribe to migration events
	// tracer/pubsub handles both Kafka and Redis Streams transparently
	callback := func(message interface{}) {
		mc.handleMigrationEvent(ctx, message)
	}

	return mc.pubsubClient.Subscribe(mc.config.TenantMigration.Topic, callback)
}

// handleMigrationEvent processes a migration event
func (mc *MigrationConsumer) handleMigrationEvent(ctx context.Context, message interface{}) {
	// Message from Redis is a string (the raw JSON from the stream)
	// Convert it to bytes for HandleMessage
	var payload []byte

	switch v := message.(type) {
	case string:
		// Redis Streams: message is already a JSON string
		payload = []byte(v)
	case []byte:
		// Kafka: message might be bytes
		payload = v
	default:
		// Fallback: marshal the message
		var err error
		payload, err = json.Marshal(message)
		if err != nil {
			mc.logger.ErrorWithTrace(ctx, err, "Failed to marshal migration event", nil)
			return
		}
	}

	log.Printf("[MIGRATION_EVENT] Received event: %s", string(payload))

	// Handle the migration event
	if err := mc.migrationSvc.HandleMessage(ctx, payload); err != nil {
		mc.logger.ErrorWithTrace(ctx, err, "Failed to handle migration event", map[string]interface{}{
			"payload": string(payload),
		})
	}
}

// StartWithRetry starts the consumer with retry logic
// Automatically reconnects on failure
func (mc *MigrationConsumer) StartWithRetry(ctx context.Context) {
	for {
		err := mc.Start(ctx)
		if err == nil || err == context.Canceled {
			return
		}

		mc.logger.ErrorWithTrace(ctx, err, "Migration consumer error, retrying in 2 seconds", nil)
		time.Sleep(2 * time.Second)
	}
}
